buffer/src/main/java/io/netty/buffer/CompositeByteBuf.java

```diff
@@ -316,43 +339,40 @@ private int addComponent0(boolean increaseWriterIndex, int cIndex, ByteBuf buffe
      * ownership of all {@link ByteBuf} objects is transfered to this {@link CompositeByteBuf}.
      */
     public CompositeByteBuf addComponents(int cIndex, ByteBuf... buffers) {
-        addComponents0(false, cIndex, buffers, 0, buffers.length);
+        checkNotNull(buffers, "buffers");
+        addComponents0(false, cIndex, buffers, 0);
         consolidateIfNeeded();
         return this;
     }
 
-    private int addComponents0(boolean increaseWriterIndex, int cIndex,
-            ByteBuf[] buffers, int offset, int endOffset) {
-        checkNotNull(buffers, "buffers");
-        int i = offset;
+    private int addComponents0(boolean increaseWriterIndex, final int cIndex, ByteBuf[] buffers, int arrOffset) {
+        final int len = buffers.length, count = len - arrOffset;
+        int ci = Integer.MAX_VALUE;
         try {
             checkComponentIndex(cIndex);
-
-            // No need for consolidation
-            while (i < endOffset) {
-                // Increment i now to prepare for the next iteration and prevent a duplicate release (addComponent0
-                // will release if an exception occurs, and we also release in the finally block here).
-                ByteBuf b = buffers[i++];
-                if (b == null) {
-                    break;
-                }
-                cIndex = addComponent0(increaseWriterIndex, cIndex, b) + 1;
-                int size = components.size();
-                if (cIndex > size) {
-                    cIndex = size;
-                }
+            shiftComps(cIndex, count); // will increase componentCount
+            ci = cIndex; // only set this after we've shifted so that finally block logic is always correct
+            int nextOffset = cIndex > 0 ? components[cIndex - 1].endOffset : 0;
+            for (ByteBuf b; arrOffset < len && (b = buffers[arrOffset]) != null; arrOffset++, ci++) {
```

Comment:
This is kind of difficult to read.   You should consider making you code slightly more verbose, but faster to consume.

Reply:
Thanks @carl-mastrangelo, I do have a habit of over-compactness sometimes. I have some smaller follow-on updates to the class in mind anyhow so will try and modify this to be a bit more verbose/readable.